### PR TITLE
PanelData: provide legacy data only when needed 

### PR DESCRIPTION
**Pull request by ryantxu**
this tries to simplify the processing pipeline by removing legacy data checks and creating legacy formats only when needed.

Since graph panel and single stat use DataFrame directly, the performance hit for `any[]` > `DataFrame[]` > `LegacyData[]` only happens for the less commonly used panel types.

TODO:  maybe in this PR, or maybe in a followup PR
- [x] Check that explore is using an angular editor before transforming to legacy
- [ ] Check that an angular editor exists in QueriesTab before transforming to legacy
- [x] Somehow preprocess all results rather than call `postProcessPanelData`

### CODE REVIEW

File: public/app/features/panel/metrics_panel_ctrl.ts [hunk 1]
```diff
@@ -140,20 +140,17 @@ class MetricsPanelCtrl extends PanelCtrl {
...
-        // The result should already be processed, but just in case
-        if (!data.legacy) {
-          data.legacy = data.series.map(v => {
-            if (isDataFrame(v)) {
-              return toLegacyResponseData(v);
-            }
-            return v;
-          });
-        }
+      if (this.useDataFrames) {
```
Comment (torkelo) [inline public/app/features/panel/metrics_panel_ctrl.ts]: missing a ! negation :)

Comment (ryantxu) [reply public/app/features/panel/metrics_panel_ctrl.ts]: I have never once made an error like this before.  My self worth is totally in question now. ;)

File: public/app/features/dashboard/panel_editor/QueriesTab.tsx [hunk 2]
```diff
@@ -75,7 +75,8 @@ export class QueriesTab extends PureComponent<Props, State> {
   }
 
   onPanelDataUpdate(data: PanelData) {
-    console.log('got data', data);
+    // Does the *panel* type matter?  I would think only
+    // the datasource type matters...
```
Comment (ryantxu) [inline public/app/features/dashboard/panel_editor/QueriesTab.tsx]: any thoughts on what we should do here?  It looks like our current approach was to check if the *panel* is angular, but I think it actually depends on if any DataSource has angular query editors.

We could cycle datasources here and check... or maybe in `onComponentUpdated` assuming the panel object changes.

File: public/app/features/dashboard/state/PanelQueryRunner.ts [hunk 3]
```diff
@@ -58,16 +51,15 @@ export class PanelQueryRunner {
 
   /**
    * Returns an observable that subscribes to the shared multi-cast subject (that reply last result).
-   * Here the caller can also control if transformations should be applied or not
    */
-  getData(format: PanelDataFormat = PanelDataFormat.Frames, applyTransforms = true): Observable<PanelData> {
-    const ensureFormats = postProcessPanelData(format);
+  getData(): Observable<PanelData> {
+    const ensureFormats = postProcessPanelData();
```
Comment (ryantxu) [inline public/app/features/dashboard/state/PanelQueryRunner.ts]: Now that `postProcessPanelData` is really the same for everything, do we need a special function?  Could this be *pre* processed before it is sent to the subject?

File: public/app/features/explore/state/reducers.ts [hunk 4]
```diff
@@ -624,7 +623,12 @@ export const processQueryResponse = (
   const logsResult = processor.getLogsResult();
 
   // For Angular editors
-  state.eventBridge.emit('data-received', legacy);
+  if (true) {
```
Comment (ryantxu) [inline public/app/features/explore/state/reducers.ts]: Is there a good way to check if the datasource has an angular editor?

**Comment by ryantxu** [comment]
@torkelo -- PR is updated, tests pass and manual testing seems OK.

In the description I point out three issues we should look at, but I'm not sure if best to merge into query-observable, or keep working on this PR.

**Comment by ryantxu** [comment]
hymmm  `lint-go` is failing... that is unrelated and not sure how to fix

File: public/app/features/dashboard/state/PanelQueryRunner.ts (hunk 3 above)
Comment (torkelo) [reply public/app/features/dashboard/state/PanelQueryRunner.ts]: yea, if it's something that always should be done I think that makes sense.

File: public/app/features/explore/state/reducers.ts (hunk 4 above)
Comment (torkelo) [reply public/app/features/explore/state/reducers.ts]: yes, if datasourceInstance.components.QueryCtrl

File: public/app/features/dashboard/state/PanelQueryRunner.ts (hunk 3 above)
Comment (ryantxu) [reply public/app/features/dashboard/state/PanelQueryRunner.ts]: thanks

File: public/app/features/explore/state/reducers.ts (hunk 4 above)
Comment (ryantxu) [reply public/app/features/explore/state/reducers.ts]: thanks

**Pull request merged**